### DISCUSSION TITLE
Reticulum "high power" LoRa node

### DISCUSSION THREAD

**Discussion by kc1awv** [markqvist/Reticulum#93] [General]
This might be a pie(raspberry pi?)-in-the-sky idea, but hear me out - what about reclaiming unused Helium miner equipment, and turning them into "high-power" Reticulum nodes?

Right now the device in the picture is set up as intended, as a LoRaWAN concentrator. The LoRa device is a RAK Wireless 2245 Pi HAT [Datasheet](https://docs.rakwireless.com/Product-Categories/WisLink/RAK2245-Pi-HAT/Datasheet/), which in the end, is just a Semtech SX1301 (2x SX125X) and a uBlox Max 7Q GPS.

As far as I can tell, the LoRa chip is SPI/I2C, and the GPS is over UART.

Some programming / development would be needed in order to attach the LoRa chip as a network interface to the underlying system (in this case, Raspbian OS), but I'm not sure where to start. There may be some prior art out there for this kind of application, so I'm tossing the idea out here for anyone else that might have an idea and see if there's any one else that might be interested.

Since the chip is a concentrator, possibly the whole package could be used like a multi-channel Reticulum router?

![IMG_0108](https://user-images.githubusercontent.com/43918257/189925152-7c8ec827-af86-4a74-918d-b335f2f3b237.jpg)

**Comment by kc1awv** [comment]
Some continuing relevant information:

[Semtech libloragw](https://github.com/Lora-net/lora_gateway) - Possible C library for radio control use

[josefmtd's kiss-lora](https://github.com/josefmtd/kiss-lora) - Prior art for attaching SX1278 as a KISS TNC, possible target to derive from to create a network interface to the OS

[markqvist's RNode Firmware](https://github.com/markqvist/RNode_Firmware) - Prior art for ATmega/ESP32 based LoRa devices, use for general guidance, possible alternative for a kissattach service - thinking of an RNode derivative, like RNode HP (high-power)

_edits may happen here if this discussion continues..._

**Comment by markqvist** [comment]
This is a very interesting proposal. That hardware would probably make a very powerful LoRa-based, multichannel base station. Best approach would *probably* be to add an SPI-based driver directly into Reticulum, and not rely on an extra kernel module or similar.

I haven't read the datasheet yet, but I am pretty certain, that if someone can get a couple of these units into my hands, I will be able to add support for them with an interface driver.

**Reply by kc1awv** [reply]
I've looked for any suppliers that may have the HAT in stock, the only ones are Adafruit and DigiKey for the 915 MHz version, and they're currently quite rare. The RAK Wireless site is sold out of the 2245 HATs.

The replacement(?) looks like the RAK 2287 HAT and a RAK 2287 mPCIe module. Still uses a SPI interface

[Datasheet - RAK 2287 HAT](https://docs.rakwireless.com/Product-Categories/WisHat/RAK2287-RAK5146-Pi-HAT/Datasheet/)
[Datasheet - RAK 2287 mPCIe Module](https://docs.rakwireless.com/Product-Categories/WisLink/RAK2287/Datasheet/)

I do have another pair of the RAK 2245 HATs coming in from Adafruit (they're not cheap!) this weekend. I'll be out on a business trip through the coming weekend, though. Worst case scenario, I could ship what I have as loaners, or provide a remote shell to a pair of devices with the radios attached.

**Reply by whitecd** [reply]
I have RAK5146-126, 915MHz, USB mPCIe modules with Listen-before-Talk (embedded SX1262 radio) and GPS.  They are based on the SX1303 Semtech GW chips.  These are the USB-based versions of the standard SPI variant.  They have a MCU managing the USB-to-SPI conversion.  Ideally, supporting both USB and SPI would be useful, since not all mPCIes support SPI. The USB version can also be used in a mPCIe-USB dongle. https://docs.rakwireless.com/Product-Categories/WisLink/RAK5146/Datasheet/#models-bundles

I could loan you one or two, if desired.

**Reply by markqvist** [reply]
That would be pretty great! I'll get back to you on this, but I have had to realise that I am juggling too many balls right now, so I'll focus of getting the next release of Reticulum, Nomadnet and Sideband out first, and then start looking at more hardware again :)

**Reply by kriloter** [reply]
> This is a very interesting proposal. That hardware would probably make a very powerful LoRa-based, multichannel base station. Best approach would _probably_ be to add an SPI-based driver directly into Reticulum, and not rely on an extra kernel module or similar.
> 
> I haven't read the datasheet yet, but I am pretty certain, that if someone can get a couple of these units into my hands, I will be able to add support for them with an interface driver.

Hello, is there anything new in this matter? I can give you a couple of these units (rak2287) if it helps. Thank you

**Reply by kriloter** [reply]
https://docs.rakwireless.com/Product-Categories/WisLink/RAK2287/Datasheet/#hardware

**Comment by kn0xyland** [comment]
Hey there, firstly I wanted to share thanks and gratitude. This is an amazing project and I'm looking forward to my supported hardware arriving and getting stuck in.

I have a Waveshare SX1303 Pi Hat I'd also love to use for a similar purpose. Will keep an eye on this thread for sure!

Hope you are all well and thanks again 🖖

**Reply by kc1awv** [reply]
Honestly, I thought this thread was dead and buried a while ago.

However, with some recent efforts towards SX126X, and RAK offering the WisBlock platform, something there might be a better target. RAK4631 is an nRF52840 with BLE and LoRa, and through WisBlock things like sensors and other interfaces could be added. Though the SX126X set of radios may be gaining support in Reticulum, I'm not sure how difficult it would be to port RNode over to ARM vs. ESP32.

For eye candy, here's something I've been working on.

![IMG_1156](https://github.com/markqvist/Reticulum/assets/43918257/86bad493-4549-4bb8-a88a-4e44d43c6f7a)

**Reply by markqvist** [reply]
That is looking really great @kc1awv! There is already a functional nRF52/ARM port, courtesy of @jacobeva. There's not a binary release of it yet, but the support is right there in the source tree and Makefile :)

**Reply by kc1awv** [reply]
Oh geez, and here I was thinking I would have a nice, quiet Easter weekend :smile:

**Reply by jacobeva** [reply]
Yes, the RAK4631 can easily be built for according to master. I have not written any driver for the ethernet port of course but that will be a fun exercise I'm sure ;)
The display will probably work utilising the existing the existing Display.h config (probably with some pins changed?), graphics won't have to be scaled because it looks similar in dimensions to the ones Mark originally intended for use on the firmware. If you would like any assistance with the RAK4631 platform for the firmware side, please DM me on matrix: @x00010:matrix.org

**Comment by chron0** [comment]
chiming in here on the PIdea - i have set up a meshtastic node on a pi that is also running an MMDVM hat - to make the most of already running hardware. I added an EBYTE SX1262 30dBm (E22-900M30S) lora module that needs spi and some gpios and that worked fine with the meshtastic linux implementation. I would guess that https://github.com/chandrawi/LoRaRF-Python may be a starting point if it's going to be python - would be a nice additional air interface option on any small linux host, pis, odroids, etc. imho.

edit: might also be interesting for running propagation nodes

**Comment by vogt31337** [comment]
@kc1awv, this looks really interesting! I recently got hold on some similar hardware and I am interested in setting up such a reticulum node with such a Lora concentrator. Was there any development until now? I would also not hesitate to develop something for reticulum, I thought about looking into this driver: https://github.com/Lora-net/sx1302_hal
It could probably support both the USB version and the SPI version, of the lora concentrator.

**Comment by tbaumann** [comment]
That would be a nice development. I'm right now just reticulum-curious. But I see it happen in other communities where mesh networks fragment into different spreading factor clusters.
A lora concentrator style chip could help bridge between those without needing a rNode on each leg.

Not to mention the high bandwith GFSK option you get too.

**Comment by k37chup** [comment]
I too have been reticulum-curious. I jumped down the Linux native Meshtastic rabbit hole and have been a little put off by its current limitations. Started looking into meshcore but it lacks a linux native options at the moment. I have played with Rnode in the past but with all the Pi Hats and USB loRa devices i have it would be awesome to be able to build out a pi reticulum node with an ebyte 900m30s or other usb lora devices. Also the SX310x concentrators have been seemingly abandoned by meshtastic since they are mostly receivers and are hard coded to a specific lora configuration. There is a sx127x in most of them that could be accessed but again doesn't seem like there is a ton of interest in development. which is a shame since there is a ton of helium equipment hitting the market and to be able to recycle the modems would be awesome.
